### PR TITLE
[RFC] Partially decouple devices from SYS_INIT

### PR DESCRIPTION
**Pull request by gmarull**
## Background

Nowadays, devices rely on the `init.h` (i.e. `SYS_INIT`) infrastructure to get initialized automatically. The `init.h` infrastructure is a basic mechanism that allows to register an init function when the system is initialized (before `main`). It provides with multiple initialization levels: `PRE_KERNEL_*`, `POST_KERNEL`, etc., and within each level a numeric priority (0-99). Using this information, each registered init entry is sorted by the linker so that the Kernel can later iterate
over them in the correct order. This all sounds nice and simple, but when it comes to devices, this mechanism has proven to be insufficient.

Before starting with the changes proposed in this patch, let's first dig into the implementation details of the current model. When devices are defined, using any of the `DEVICE_*DEFINE` macros, they also create an init entry using the internal `init.h` APIs (see `Z_DEVICE_INIT_ENTRY_DEFINE`). This entry, stores a pointer to the device init call and to the device itself. As the reader can imagine, this implies a coupling between `init.h` and `device.h`. The only link between a device and an init entry is the device pointer stored in the init entry. This allows the Kernel init machinery to call the init
function with the right device pointer. However, there is no direct relationship between a device and its init function, that is, `struct device` does not keep the device init function reference. This is not a problem nowadays, but it could be a problem if features like deferred initialization or init/de-init have to be implemented. However, in reality, this is a _secondary_ problem. The most problematic issue we have today is that devices are mixed with `SYS_INIT` calls. They are all part of the same init block, and are treated equally. So for example, one can theoretically have a system where the init sequence can be like:

```
- PRE_KERNEL_1
  - init call 1
  - device 0
  - init call 2
  - init call 3
  - device 1
  - device 2
  - init call 4
- PRE_KERNEL_2
  - init call 5
  - device 3
  ...
```

This is problematic because:

(1) Init calls can depend on devices, but dependency is not tracked anywhere. So the user must check that init priorities are correct to avoid runtime failures.
(2) Device drivers can have multiple instances, and each instance may require a different set of priorities, while `SYS_INIT` calls are singletons.
(3) Devices don't likely need so many init priorities (`SMP`?  `APPLICATION`?)

(1) is particularly important because init calls do not have a specific purpose. They usage ranges from SoC init code, to system services. So it's _unpredictable_ what can happen in there. (2) is a tangential topic, actually not fixed by this patch, even though it helps. (3) is more of a post-patch cleanup we need.

## So, what does this patch propose...?

**First, this patch is still a HACK, so please, focus on the description rather than with the implementation**

So it's about providing devices with their own init infrastructure,
minimizing the coupling with init.h. This patch groups devices in a
separate section, but, keeps the same init levels as before. Therefore, the list
above would look like:

```
/* devices */
- PRE_KERNEL_1
  - device 0
  - device 1
  - device 2
- PRE_KERNEL_2
  - device 3
  ...

/* init calls */
- PRE_KERNEL_1
  - init call 1
  - init call 2
  - init call 3
  - init call 4
- PRE_KERNEL_2
  - init call 5
  ...
```

This means that **it is no longer possible** to mix init calls and devices within the same level. So how does it work now? Within each level, devices are initialized first, then init calls. It is done this way because I believe it is init calls who depend on devices and not vice versa. I may be wrong, and so the whole proposal would need a rework. So the init order would look like:

```
- PRE_KERNEL_1
  - device 0
  - device 1
  - device 2
  - init call 1
  - init call 2
  - init call 3
  - init call 4
- PRE_KERNEL_2
  - device 3
  - init call 5
  ...
```

## Why does this matter for future development?

First, we have devices grouped at least on a level basis. This means that we could likely start using devicetree ordinals, by reducing the source of problems to the init level only. We can also re-think device init levels (probably init levels as well, hey `PRE_KERNEL_1/2`). For example, there could be pre-kernel and post-kernel devices only. The other side effect of this change is that `struct device` stores the device init call, so we can potentially do things like defer device initialization or implement things like init/de-init. They all come with their own complexity, of course, but this patch could be a step forward.

## Problems

This is a breaking change. Sorry, guys, another one to the list. The API does not change, so builds should continue to work. However, the system changes its behavior, so if anyone was relying on mixed init call/device sequences, things will break. This is why it is important to analyze lots of use cases before moving forward with this proposal.

### CODE REVIEW

**Comment by gmarull** [comment]
Looks like removal of init.h from device.h triggers a new include party

ref.
https://github.com/zephyrproject-rtos/zephyr/pull/62005
https://github.com/zephyrproject-rtos/zephyr/pull/62007
https://github.com/zephyrproject-rtos/zephyr/pull/62008

**Comment by gmarull** [comment]
> > I may be wrong, and so the whole proposal would need a rework. So the init order would look like:
> > ```
> > - PRE_KERNEL_1
> >   - device 0
> >   - device 1
> >   - device 2
> >   - init call 1
> >   - init call 2
> >   - init call 3
> >   - init call 4
> > - PRE_KERNEL_2
> >   - device 3
> >   - init call 5
> >   ...
> > ```
> 
> We have many platforms that initialize all kind of things platform related very early, probably infra that would be needed for device drivers, i.e. memory, power management, clocks and so on, will this work? For example, randomly picked soc:
> 
> `soc/arm/nxp_kinetis/k6x/soc.c` which calls `SYS_INIT(k6x_init, PRE_KERNEL_1, 0);`

These should likely be moved to `EARLY`, or better, introduce a new SoC hook that is guaranteed to be called _early_.

**Comment by gmarull** [comment]
@fabiobaltieri these changes break the check_init_priorities script because this disappears from ELF object files:

```
Relocation section '.rel.z_init_PRE_KERNEL_140_00104_' at offset 0x6b30 contains 2 entries:
 Offset     Info    Type            Sym.Value  Sym. Name
00000000  00001d02 R_ARM_ABS32       00000001   gpio_nrfx_init
00000004  00005a02 R_ARM_ABS32       00000000   __device_dts_ord_104
```

I haven't been able to find an alternative for now.

**Comment by gmarull** [comment]
@fabiobaltieri ok, found a solution for the check_init_priorities.py script (thanks for the hints). It turns out to be simpler now.

File: scripts/build/check_init_priorities.py [hunk 1]
```diff
@@ -120,60 +121,26 @@ class variable.
...
-
-        _, device_ord = sym_name.split(_DEVICE_ORD_PREFIX)
-        return int(device_ord)
-
     def _find_defined_devices(self):
-        """Find the device structures defined in the object file."""
-        self.defined_devices = {}
+        dev_syms = []
+        rel_sects = {}
+        for i, section in enumerate(self._elf.iter_sections()):
```
Comment (fabiobaltieri) [inline scripts/build/check_init_priorities.py:129]: Would it work to use `enumerate(self._elf.iter_sections(), 1)` and drop the `+ 1` in line 142?

File: scripts/build/check_init_priorities.py [hunk 2]
```diff
@@ -32,9 +32,9 @@
                                 "dts", "python-devicetree", "src"))
 from devicetree import edtlib  # pylint: disable=unused-import
 
-# Prefix used for relocation sections containing initialization data, as in
-# sequence of "struct init_entry".
-_INIT_SECTION_PREFIX = (".rel.z_init_", ".rela.z_init_")
+# Prefix used for relocation sections containing devices, as in sequence of
+# "struct device".
+_DEVICE_SECTION_PREFIX = (".rel._device.static.", ".rela._device.static.")
```
[the remark this answers is not in the document]
Comment (gmarull) [reply scripts/build/check_init_priorities.py]: removed

File: scripts/build/check_init_priorities.py (hunk 1 above)
Comment (gmarull) [reply scripts/build/check_init_priorities.py]: added, was not aware of this feature of enumerate...

Comment (gmarull) [reply scripts/build/check_init_priorities.py:127]: actually this was wrong, posted a new revision

**Comment by carlescufi** [comment]
Arch WG:

- Description of the proposal by @gmarull 
- A [script](https://github.com/zephyrproject-rtos/zephyr/blob/main/scripts/build/check_init_priorities.py) already verifies that the DT ordinal order is respected with init prios
-  It's a breaking change due to the fact that it will change the order of some initializations
- @nashif mentions that we need to figure out which hooks we need and we don't and which levels before merging anything. Ideally we would map out the whole sequence of changes to find out if we do everything at once (or at least within a release cycle)
- @nashif also asks if it's possible to link device inits to subsystem init calls. This was tried by @JordanYates but ultimately rejected
- @henrikbrixandersen likes the approach, mentions a [PR by @nashif](https://github.com/zephyrproject-rtos/zephyr/pull/61547/files). @henrikbrixandersen also mentions moving away from `SYS_INIT` to services for subsystems, where you can obtain a reference to a service and whether it's been initialized or not, etc.
- @gmarull agrees that we need to fix the SoC init problem first. Also, avoid modifying `SYS_INIT`

Sequence proposed:
1. Clean up the SoC init code before 3.5.0
2. Potentially clean up other things (eg. init levels)
3. Prepare a branch with a full PoC now
4. Merge the branch after 3.5.0

**Comment by nashif** [comment]
some thoughts re init levels:
- `SMP` is used by a couple of items, maybe we should just rename it to LATE, so we have EARLY and LATE and it can be used by smp init code but other things as well.
- Look into merging `PRE_KERNEL_1` and `PRE_KERNEL_2`
- Introduce new `PLATFORM` init level and move all SYS_INIT calls (soc_init and friends) with PRE_KERNEL_ priority level 0 into it.

Not sure if we can deprecate some of existing levels while we introduce new ones, would be great (maybe leave them in place while we add the new ones...)

**Comment by quic-egmc** [comment]
This is dancing around the fundamental issue that we are not directly stating dependencies but using init phase/level to infer a dependency.  If we're going to fix something let's fix that.

There is a model used by the UEFI folks that is interesting where a [dependency expression](https://uefi.org/specs/PI/1.8/V1_PEI_Foundation.html#dependency-expressions) states what you need to be present (typically protocol GUIDs indicating the availability of services because it's UEFI and everything is GUIDs) before you can be loaded.  In UEFI the depex is compiled into a bytecode but evaluated at runtime.  If we wanted to do this the "Zephyr way" we could do something simpler and evaluate at build time.

**Comment by gmarull** [comment]
> This is dancing around the fundamental issue that we are not directly stating dependencies but using init phase/level to infer a dependency. If we're going to fix something let's fix that.
> 
> There is a model used by the UEFI folks that is interesting where a [dependency expression](https://uefi.org/specs/PI/1.8/V1_PEI_Foundation.html#dependency-expressions) states what you need to be present (typically protocol GUIDs indicating the availability of services because it's UEFI and everything is GUIDs) before you can be loaded. In UEFI the depex is compiled into a bytecode but evaluated at runtime. If we wanted to do this the "Zephyr way" we could do something simpler and evaluate at build time.

Note that device dependencies are tracked (using ordinals), but they are not used yet. This patch assumes that there are no in-level dependencies from devices to init calls. I think it's a safe assumption (and future limitation) that makes things much easier, provided some cases like SoC init calls are moved to a very early init stage. If we need interdependencies with init calls, then a mechanism like that could be useful, given it can be done at compile time. I hope that the new collab branch will help on analyzing if this proposal works in practice and not just with some limited tests.

**Comment by dleach02** [comment]
It seems like a nice capability about the current initialization flow is the amount of flexibility one has in ordering init calls with device inits. I think we should be careful in assuming that there are not going to be situations where maybe you do need to have a device initialize with a related init call between the next device all within the same level. The current system can do this now by setting the level and priority appropriately. 

But if you want to make it a limitation I think we should be careful in deciding this to solve what appears to be a dependency issue.

**Comment by benediktibk** [comment]
> It seems like a nice capability about the current initialization flow is the amount of flexibility one has in ordering init calls with device inits. I think we should be careful in assuming that there are not going to be situations where maybe you do need to have a device initialize with a related init call between the next device all within the same level. The current system can do this now by setting the level and priority appropriately.
> 
> But if you want to make it a limitation I think we should be careful in deciding this to solve what appears to be a dependency issue.

Wouldn't it be possible to solve this with the new approach as well? I think about creating a custom device tree binding for this init call where the instance has a dependency on the device it should be initialized after, and the device which should be initialized after this custom device with a dependency on this custom device.

To clarify it: 
A: device 1
B: device 2
C: custom device

dependency graph: A <- C <- B

**Comment by gmarull** [comment]
> Wouldn't it be possible to solve this with the new approach as well? I think about creating a custom device tree binding for this init call where the instance has a dependency on the device it should be initialized after, and the device which should be initialized after this custom device with a dependency on this custom device.
> 
> To clarify it: A: device 1 B: device 2 C: custom device
> 
> dependency graph: A <- C <- B

Devicetree bindings are the schema for devicetree, and devicetree is just about hardware description, so I don't think this approach makes sense.

**Comment by benediktibk** [comment]
> > Wouldn't it be possible to solve this with the new approach as well? I think about creating a custom device tree binding for this init call where the instance has a dependency on the device it should be initialized after, and the device which should be initialized after this custom device with a dependency on this custom device.
> > To clarify it: A: device 1 B: device 2 C: custom device
> > dependency graph: A <- C <- B
> 
> Devicetree bindings are the schema for devicetree, and devicetree is just about hardware description, so I don't think this approach makes sense.

It then really depends on how much you want to put into the devicetree and what you consider as a device.

- An ethernet device? For sure.
- A http web server? Probably not.
- An IP stack? Maybe?

Just as a random example where in my opinion there are certain grey areas what can and should be a device and what not.

**Comment by gmarull** [comment]
> It seems like a nice capability about the current initialization flow is the amount of flexibility one has in ordering init calls with device inits. I think we should be careful in assuming that there are not going to be situations where maybe you do need to have a device initialize with a related init call between the next device all within the same level. The current system can do this now by setting the level and priority appropriately.
> 

Well, the current system can achieve that, but it's also not solving cases where different device instances require different priorities (see a board that can't use the current model as a result https://github.com/zephyrproject-rtos/zephyr/pull/49474). To me, the current flexibility is just nice on the paper, but in practice, it's proven to result in a complicated and problematic design.

> But if you want to make it a limitation I think we should be careful in deciding this to solve what appears to be a dependency issue.

It is a dependency issue. But, if we find that most if not all cases can be solved by the following:

1. Convert some code to devices (we have quite a few "fake driver code" in SYS_INITs)
2. Introduce some new hooks (e.g. platform init)
3. Move SYS_INITs after devices

Then things become simpler. On other systems like Linux, init calls don't even have a priority, so there are definitely many ways to design this.

**Comment by gmarull** [comment]
> > > Wouldn't it be possible to solve this with the new approach as well? I think about creating a custom device tree binding for this init call where the instance has a dependency on the device it should be initialized after, and the device which should be initialized after this custom device with a dependency on this custom device.
> > > To clarify it: A: device 1 B: device 2 C: custom device
> > > dependency graph: A <- C <- B
> > 
> > 
> > Devicetree bindings are the schema for devicetree, and devicetree is just about hardware description, so I don't think this approach makes sense.
> 
> It then really depends on how much you want to put into the devicetree and what you consider as a device.
> 
> * An ethernet device? For sure.
> * A http web server? Probably not.
> * An IP stack? Maybe?

How is an IP stack hardware related? It's a _software stack_. I think devicetree is quite explicit on what one should expect in there; however, Zephyr has abused on devicetree because it's just "practical" and we have nothing better. I'd personally not have crossed that boundary because deviating from a spec we do not have much influence on will be problematic in the future. Also, devicetree is not really well maintained right now, so things will just become worse over time if its original purpose is blurred.

**Comment by fabiobaltieri** [comment]
> Wouldn't it be possible to solve this with the new approach as well? I think about creating a custom device tree binding for this init call where the instance has a dependency on the device it should be initialized after, and the device which should be initialized after this custom device with a dependency on this custom device.
> 
> To clarify it: A: device 1 B: device 2 C: custom device
> 
> dependency graph: A <- C <- B

If there's such a dependency you could make the case that C should indeed be represented as a device, so I think it make sense that it would have corresponding device structure (how else would you point B to it?). Then the current system works, if you set everything at the same level+priority then ordinals would do the rest.

**Comment by dleach02** [comment]
I look at the initialization framework as just an initialization framework that happens to be used by devices and SYS_Init. This is rather elegant and clean solution. I don't have a direct example of where intertwining device and SYS_Init need to occur but with this proposal we are explicitly calling out different initialization frameworks that happen in different timelines.

I'm looking at the reference PR you highlighted but I can't find the SHA in the note about the use case that can't be handled? 

Since we keep calling it a dependency issue then we should focus on how to express dependency without introducing artificial limitations on part of the initialization framework. Device tree expression of dependency is a good start but also pulling in SYS_Init() into this dependency tree would retain the current functionality and flexibility.

**Comment by GeorgeCGV** [comment]
> This means that it is no longer possible to mix init calls and devices within the same level. So how does it work now? Within each level, devices are initialized first, then init calls. It is done this way because I believe it is init calls who depend on devices and not vice versa.

@gmarull I have cases where sys init has to perform a set of actions (gpio, pwm, etc...) before a device can be initialized. Therefore, when that is no longer allowed, such code would require a device driver. That results in a bigger footprint, apart from the code not being a device.

**Comment by gmarull** [comment]
> > This means that it is no longer possible to mix init calls and devices within the same level. So how does it work now? Within each level, devices are initialized first, then init calls. It is done this way because I believe it is init calls who depend on devices and not vice versa.
> 
> @gmarull I have cases where sys init has to perform a set of actions (gpio, pwm, etc...) before a device can be initialized. Therefore, when that is no longer allowed, such code would require a device driver. That results in a bigger footprint, apart from the code not being a device.

Being more specific would help. Many of these cases often exist because of a bad DT description, device drivers doing too much, some driver features did not exist (e.g. regulators or gpio hogs), etc.

**Comment by benediktibk** [comment]
> > > This means that it is no longer possible to mix init calls and devices within the same level. So how does it work now? Within each level, devices are initialized first, then init calls. It is done this way because I believe it is init calls who depend on devices and not vice versa.
> > 
> > 
> > @gmarull I have cases where sys init has to perform a set of actions (gpio, pwm, etc...) before a device can be initialized. Therefore, when that is no longer allowed, such code would require a device driver. That results in a bigger footprint, apart from the code not being a device.
> 
> Being more specific would help. Many of these cases often exist because of a bad DT description, device drivers doing too much, some driver features did not exist (e.g. regulators or gpio hogs), etc.

I can think of an example, or rather actually have one available: A hardware with two devices (GPIO expanders, to be specific) where it is necessary to execute a reset via a dedicated GPIO. So far, so good, each device can trigger this by itself and get the corresponding GPIO via devicetree. But in this specific case the hardware designer decided to save a GPIO and connect both reset signals to the same output on the MCU. Therefore, one needs to execute a dedicated reset on this GPIO before the device gets initialized.

I personally do not have a big issue with using a dedicated reset-device for this purpose, as it is available already anyway for other reasons. But I assume some people might consider this a waste of resources.

**Comment by gmarull** [comment]
> > > > This means that it is no longer possible to mix init calls and devices within the same level. So how does it work now? Within each level, devices are initialized first, then init calls. It is done this way because I believe it is init calls who depend on devices and not vice versa.
> > > 
> > > 
> > > @gmarull I have cases where sys init has to perform a set of actions (gpio, pwm, etc...) before a device can be initialized. Therefore, when that is no longer allowed, such code would require a device driver. That results in a bigger footprint, apart from the code not being a device.
> > 
> > 
> > Being more specific would help. Many of these cases often exist because of a bad DT description, device drivers doing too much, some driver features did not exist (e.g. regulators or gpio hogs), etc.
> 
> I can think of an example, or rather actually have one available: A hardware with two devices (GPIO expanders, to be specific) where it is necessary to execute a reset via a dedicated GPIO. So far, so good, each device can trigger this by itself and get the corresponding GPIO via devicetree. But in this specific case the hardware designer decided to save a GPIO and connect both reset signals to the same output on the MCU. Therefore, one needs to execute a dedicated reset on this GPIO before the device gets initialized.

You can likely solve this by just defining the reset pin to just one of the expanders, right? Considering both can have the same init order, so their exact order not predictable, you would likely need to define some shared state between instances in the driver code (a static flag that the first one to reset sets).

> 
> I personally do not have a big issue with using a dedicated reset-device for this purpose, as it is available already anyway for other reasons. But I assume some people might consider this a waste of resources.

**Comment by henrikbrixandersen** [comment]
@benediktibk From your description, this sounds like a good candidate for some board level init code.

**Comment by gmarull** [comment]
> @benediktibk From your description, this sounds like a good candidate for some board level init code.

That would mean using device-specific properties (`reset-gpios` or whatever property) in board init code. I think the right solution would be to modify the driver so that it supports such functionality (as described above, in the form of an instance-shared flag). I find most board init code a sign of missing functionality or flexibility in the current model. We could have gone the _manual_ init procedure (decided at soc, board code) as many systems do, that while more cumbersome, can likely solve all edge cases and is consistent (ie doesn't mix things).

**Comment by fabiobaltieri** [comment]
You can still sneak sneak code between devices that have different levels, although having an actual device backing the code would guarantee that it gets run in the right spot in the sequence, may as well be worth the extra 20 to 28 bytes of flash.

**Comment by GeorgeCGV** [comment]
@gmarull for example a device might require a negative voltage. That can be achieved using an external IC. That is where the driver model fits. But, sometimes, MCU's PWM is enough. In that case a `SYS_INIT` function that runs before the device is a perfect solution.

**Comment by gmarull** [comment]
> @gmarull for example a device might require a negative voltage. That can be achieved using an external IC. That is where the driver model fits. But, sometimes, MCU's PWMs can be used for that purpose. In that case a `SYS_INIT` function that runs before the device is a perfect solution.

I think the response to that is: implement a regulator driver, and make your device dependent on it. Devicetree will capture this correctly (remember that what you describe is not captured in DT) and things should _just work_. It's definitely more ROM/RAM, we can't argue against that, but I see no better way in the current model. But since Zephyr runs in very constrained systems, I assume we'll end up needing the pre+post-device sys_init, so you can still do things like this out of the device model.

**Comment by henrikbrixandersen** [comment]
> I think the right solution would be to modify the driver so that it supports such functionality

Sure, but when doing products based on a specific Zephyr release using upstream drivers you sometimes have to choose whether to modify the upstream driver to extend support for exactly your board configuration and carry a local patch - or circumvent the limitations by using board-specific initialisation code. Both scenarios are valid solutions.

**Comment by gmarull** [comment]
> > I think the right solution would be to modify the driver so that it supports such functionality
> 
> Sure, but when doing products based on a specific Zephyr release using upstream drivers you sometimes have to choose whether to modify the upstream driver to extend support for exactly your board configuration and carry a local patch - or circumvent the limitations by using board-specific initialisation code. Both scenarios are valid solutions.

Then we just need the pre+post device sys_init capability, which, at this point, I think it's unavoidable if we want to move forward. However, I'd be more skeptical about allowing such board init code upstream.

**Comment by henrikbrixandersen** [comment]
> Then we just need the pre+post device sys_init capability, which, at this point, I think it's unavoidable if we want to move forward.

That's not a bad idea.

> However, I'd be more skeptical about allowing such board init code upstream.

Agreed.

**Comment by fabiobaltieri** [comment]
> > Then we just need the pre+post device sys_init capability, which, at this point, I think it's unavoidable if we want to move forward.
> 
> That's not a bad idea.

Well, in this case "post" is the level and "pre" is the previous level, not much to invent. :-)

**Comment by gmarull** [comment]
> > > Then we just need the pre+post device sys_init capability, which, at this point, I think it's unavoidable if we want to move forward.
> > 
> > 
> > That's not a bad idea.
> 
> Well, in this case "post" is the level and "pre" is the previous level, not much to invent. :-)

we won't always have a previous level, what if in PRE_KERNEL_1, and you need something before a device is initialized?

**Comment by henrikbrixandersen** [comment]
> Well, in this case "post" is the level and "pre" is the previous level, not much to invent. :-)

Except that the "pre" may depend on devices initialized in the previous level, and "post" may depend on devices initialized in the current level and so forth.

**Comment by fabiobaltieri** [comment]
> Except that the "pre" may depend on devices initialized in the previous level, and "post" may depend on devices initialized in the current level and so forth.

Wouldn't we always do first devices then sys_init with this proposal? so "pre" would always be after the devices of the previous level and "post" would always be after the devices of the current level, unless I missed something... the cycle is

```c
static void z_sys_init_run_level(enum init_level level)
{
...
	for (const struct device *dev = dev_levels[level]; dev < dev_levels[level+1]; dev++) {
	...
	}
...
 	for (const struct init_entry *entry = levels[level]; entry < levels[level+1]; entry++) {
	...
	}
};
```

**Comment by gmarull** [comment]
> > we won't always have a previous level, what if in PRE_KERNEL_1, and you need something before a device is initialized?
> 
> EARLY? There are no devices there right?

right, it's an option now because devices are not allowed there.

**Comment by gmarull** [comment]
@fabiobaltieri another point: you can't always use the level before: what if you need the Kernel to be up and running? you need something after PRE_KERNEL and before POST_KERNEL devices.

**Comment by benediktibk** [comment]
If I understand it correctly, the basic problem is the mixture of init calls which either have device or do not have one and their correct dependencies. So far this was resolved via init levels and priorities, which would in the end want to get rid of.
What about some sort of SYS_INIT, which does not have the overhead of a complete device, but can depend on device(s) and vice versa?

**Comment by JordanYates** [comment]
> If I understand it correctly, the basic problem is the mixture of init calls which either have device or do not have one and their correct dependencies. So far this was resolved via init levels and priorities, which you would want to get rid of in the end. What about some sort of SYS_INIT, which does not have the overhead of a complete device, but can depend on device(s) and vice versa?

`SYS_INIT` already does not have the overhead of a complete device. The ability to add dependencies between `SYS_INIT` and devices was proposed in #49318 and rejected.

**Comment by henrikbrixandersen** [comment]
We've briefly discussed introducing a "services" concept parallel to `SYS_INIT` where a service can depend on devices and other services, but I don't have anything concrete yet.

**Comment by gmarull** [comment]
> If I understand it correctly, the basic problem is the mixture of init calls which either have device or do not have one and their correct dependencies. So far this was resolved via init levels and priorities, which would in the end want to get rid of. What about some sort of SYS_INIT, which does not have the overhead of a complete device, but can depend on device(s) and vice versa?

First, the overhead of a device is not so high (look at struct device), so having 2 or 3 extra devices won't make a big difference in most cases. Having 2 entities using the same ordinal set would also require some coordination (likely some extra ROM/RAM?). The main problem with SYS_INITs is that dependencies are inferred from devicetree, and there's no way to describe in DT such SYS_INIT functions. One solution would be to be able to add SYS_INIT dependencies to devices, but we do not have the compile-time capability to do that easily. And to me, elf parsing/hacking is a no-go, better accept the reality than go into such nonsense.

**Pull request closed without merging**